### ISSUE TITLE
google_service_account_iam_member can cause "Root object was present, but now absent"

### ISSUE THREAD

**Issue by jbridger** [hashicorp/terraform-provider-google#19680]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to a user, that user is claiming responsibility for the issue.
* Customers working with a Google Technical Account Manager or Customer Engineer can ask them to [reach out internally](https://github.com/hashicorp/terraform-provider-google/wiki/Customer-Contact#raising-gcp-internal-issues-with-the-provider-development-team) to expedite investigation and resolution of this issue.

### Terraform Version & Provider Version(s)

Terraform v1.9.6
on darwin_arm64 and linux_amd64
+ provider registry.terraform.io/hashicorp/google v6.4.0
+ provider registry.terraform.io/hashicorp/google-beta v6.4.0

### Affected Resource(s)

google_service_account_iam_member

### Terraform Configuration

```tf
resource "google_project_iam_member" "gcp_support" {
  member  = "user:blah@xwf.google.com"
  role    = "roles/bigquery.dataViewer"
  project = local.project.id
}
```

### Debug Output

```
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to
│ google_project_iam_member.gcp_support, provider
│ "provider[\"registry.terraform.io/hashicorp/google\"]" produced an
│ unexpected new value: Root object was present, but now absent.
│ 
│ This is a bug in the provider, which should be reported in the provider's
│ own issue tracker.
``` 

### Expected Behavior

The IAM member should be created.

### Actual Behavior

It fails to apply and outputs an error that doesn't help with identifying the cause.

### Steps to reproduce

1. `terraform apply`

### Important Factoids

I am using Terraform to add a principal from Google's Extended Workforce (xWF) to provide access to BigQuery. This principal has the domain `xwf.google.com`.

If I add this principal via the BigQuery web console in GCP, the principal is automatically converted from `blah@xwf.google.com` to `blah@google.com`.

I'm assuming that this why the provider fails to apply the resource if you have the xwf subdomain. It works if I change my terraform resource to use the `blah@google.com` principal.

**Comment by tpdownes** [comment]
@jbridger you might rename the issue to read `google_project_iam_member`

**Comment by slevenick** [comment]
Hmmm, can you share the debug logs for the apply that fails? I'm wondering if this is a provider-side issue or an API-side issue.

> If I add this principal via the BigQuery web console in GCP, the principal is automatically converted from blah@xwf.google.com to blah@google.com.

This indicates to me that it may be an API-side issue. If the API converts xwf.google.com -> google.com then it appears the API doesn't respect the prefix. We won't be able to do much on the Terraform side if that's the case

You can share the logs with my @google email address if they contain sensitive information.

**Comment by jbridger** [comment]
> Hmmm, can you share the debug logs for the apply that fails? I'm wondering if this is a provider-side issue or an API-side issue.
> 
> > If I add this principal via the BigQuery web console in GCP, the principal is automatically converted from [blah@xwf.google.com](mailto:blah@xwf.google.com) to [blah@google.com](mailto:blah@google.com).
> 
> This indicates to me that it may be an API-side issue. If the API converts xwf.google.com -> google.com then it appears the API doesn't respect the prefix. We won't be able to do much on the Terraform side if that's the case
> 
> You can share the logs with my @google email address if they contain sensitive information.

I'll get the debug logs and send them to you soon. Thanks

**Comment by slevenick** [comment]
> > Hmmm, can you share the debug logs for the apply that fails? I'm wondering if this is a provider-side issue or an API-side issue.
> > > If I add this principal via the BigQuery web console in GCP, the principal is automatically converted from [blah@xwf.google.com](mailto:blah@xwf.google.com) to [blah@google.com](mailto:blah@google.com).
> > 
> > 
> > This indicates to me that it may be an API-side issue. If the API converts xwf.google.com -> google.com then it appears the API doesn't respect the prefix. We won't be able to do much on the Terraform side if that's the case
> > You can share the logs with my @google email address if they contain sensitive information.
> 
> I'll get the debug logs and send them to you soon. Thanks

Yeah, it looks like from the logs that the API is making the change. Unfortunately I think that means we can't do anything here.

We don't want to be in the business of comparing emails to equivalency because there are going to be a million rules that we don't have access to for comparison. Apparently the API thinks that xwf.google.com and .google.com are equivalent, but we have no way to know that and don't want to hard-code rules like that into the provider to help with comparison.

Are xwf.google.com and google.com equivalent for your use case? Like, if you grant permission to the .google.com address does it work correctly? I would recommend just using .google.com in your TF file to fix this issue

**Comment by jbridger** [comment]
When I discovered that if you add the user via the GCP Web Console it converted the address from `xwf.google.com` to `google.com`, I updated my Terraform resource to use the latter address which works just fine.

It just took me a while to realize that was the cause. Is it possible to adjust the error message in Terraform to feedback to the user that something like this is happening?

**Comment by slevenick** [comment]
> When I discovered that if you add the user via the GCP Web Console it converted the address from `xwf.google.com` to `google.com`, I updated my Terraform resource to use the latter address which works just fine.
> 
> It just took me a while to realize that was the cause. Is it possible to adjust the error message in Terraform to feedback to the user that something like this is happening?

Unfortunately not really. The error message is coming from Terraform core when it fails to read the resource that it thinks it just created. We can't modify the error coming out of there. We could only add a log message at create time, but it would just go to the debug logs and wouldn't create anything user-facing for the most part
